driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
+                                    return;
+                                }
+
+                                String currentKeyspace = connection.keyspace();
+                                String prepareKeyspace = toPrepare.getQueryKeyspace();
+                                if (prepareKeyspace != null && (currentKeyspace == null || !currentKeyspace.equals(prepareKeyspace))) {
+                                    // This shouldn't happen in normal use, because a user shouldn't try to execute
+                                    // a prepared statement with the wrong keyspace set.
+                                    // Fail fast (we can't change the keyspace to reprepare, because we're using a pooled connection
+                                    // that's shared with other requests).
+                                    release(connection);
+                                    throw new IllegalStateException(String.format("Statement was prepared on keyspace %s, can't execute it on %s (%s)",
+                                        toPrepare.getQueryKeyspace(), connection.keyspace(), toPrepare.getQueryString()));
+                                }
+
+                                logger.info("Query {} is not prepared on {}, preparing before retrying executing. "
+                                        + "Seeing this message a few times is fine, but seeing it a lot may be source of performance problems",
+                                    toPrepare.getQueryString(), connection.address);
+
+                                write(connection, prepareAndRetry(toPrepare.getQueryString()));
+                                // we're done for now, the prepareAndRetry callback will handle the rest
+                                return;
+                            default:
+                                release(connection);
+                                if (metricsEnabled())
+                                    metrics().getErrorMetrics().getOthers().inc();
+                                break;
+                        }
+
+                        if (retry == null)
+                            setFinalResult(connection, response);
+                        else {
+                            switch (retry.getType()) {
+                                case RETRY:
+                                    ++retriesByPolicy;
+                                    if (logger.isDebugEnabled())
+                                        logger.debug("Doing retry {} for query {} at consistency {}", retriesByPolicy, statement, retry.getRetryConsistencyLevel());
+                                    if (metricsEnabled())
+                                        metrics().getErrorMetrics().getRetries().inc();
+                                    retry(true, retry.getRetryConsistencyLevel());
+                                    break;
+                                case RETHROW:
+                                    setFinalResult(connection, response);
```

Comment:
That sounds good to me, thanks for clearing that up.  So a user could actually still see a ReadTimeoutException if speculative executions are in progress, but I can see why you wouldn't want to wait for the other queries to complete as you could see the same thing later.   The speculative execution time I imagine is going to typically be very low (milliseconds) which is should be less than the cassandra read timeout by some large factor.